File: compress/src/java/compresslib/Containers.java
package compresslib;

import java.util.Arrays;

/**
 * zlib 과 gzip 컨테이너 — SPEC §10.8 (RFC 1950, RFC 1952).
 *
 * <p>gzip 머리의 MTIME 을 <b>0 으로 못 박는다.</b> 진짜 gzip 은 파일의
 * 수정 시각을 적어서 같은 입력에 같은 바이트가 안 나온다 — 재현이 안
 * 된다.
 */
public final class Containers {
  public static final int ZLIB_CMF = 0x78;  // CM 8 = deflate, CINFO 7
  // (CMF<<8|FLG) 가 31 의 배수여야 한다
  public static final int ZLIB_FLG = 0x9C;
  public static final int GZIP_DEFLATE = 8;
  public static final int GZIP_OS_UNKNOWN = 255;

  private Containers() {}

  public static byte[] zlibCompress(byte[] src) {
    ByteBuf out = new ByteBuf();
    out.push(ZLIB_CMF);
    out.push(ZLIB_FLG);
    out.extend(Deflate.deflateRaw(src));
    int a = Checksums.adler32(src);
    for (int i = 3; i >= 0; i--) {
      out.push((a >>> (8 * i)) & 0xFF);
    }
    return out.bytes();
  }

  public static byte[] zlibDecompress(byte[] src) {
    if (src.length < 6) {
      throw new CodecException("zlib 스트림이 너무 짧다");
    }
    int cmf = src[0] & 0xFF;
    int flg = src[1] & 0xFF;
    if ((cmf & 0x0F) != 8) {
      throw CodecException.of("zlib CM 이 8 이 아니다: %d", cmf & 0x0F);
    }
    if (((cmf << 8) | flg) % 31 != 0) {
      throw new CodecException(
          "zlib 머리의 검사식이 31 로 안 나눠진다");
    }
    if ((flg & 0x20) != 0) {
      throw new CodecException(
          "미리 정한 사전(FDICT)은 지원하지 않는다");
    }
    byte[] body = Arrays.copyOfRange(src, 2, src.length - 4);
    byte[] out = Inflate.inflateRaw(body);
    int want = 0;
    for (int i = src.length - 4; i < src.length; i++) {
      want = (want << 8) | (src[i] & 0xFF);
    }
    if (Checksums.adler32(out) != want) {
      throw new CodecException("Adler-32 가 다르다");
    }
    return out;
  }

  public static byte[] gzipCompress(byte[] src) {
    ByteBuf out = new ByteBuf();
    int[] head = {0x1F, 0x8B, GZIP_DEFLATE, 0,
        0, 0, 0, 0, 0, GZIP_OS_UNKNOWN};
    for (int b : head) {
      out.push(b);
    }
    out.extend(Deflate.deflateRaw(src));
    int c = Checksums.crc32(src);
    int n = src.length;
    for (int i = 0; i < 4; i++) {
      out.push((c >>> (8 * i)) & 0xFF);
    }
    for (int i = 0; i < 4; i++) {
      out.push((n >>> (8 * i)) & 0xFF);
    }
    return out.bytes();
  }

  public static byte[] gzipDecompress(byte[] src) {
    if (src.length < 18) {
      throw new CodecException("gzip 스트림이 너무 짧다");
    }
    if ((src[0] & 0xFF) != 0x1F || (src[1] & 0xFF) != 0x8B) {
      throw new CodecException("gzip 매직이 아니다");
    }
    if ((src[2] & 0xFF) != GZIP_DEFLATE) {
      throw CodecException.of("gzip CM 이 8 이 아니다: %d",
          src[2] & 0xFF);
    }
    int flg = src[3] & 0xFF;
    int pos = 10;
    if ((flg & 0x04) != 0) {      // FEXTRA
      pos += 2 + ((src[pos] & 0xFF) | ((src[pos + 1] & 0xFF) << 8));
    }
    for (int bit : new int[] {0x08, 0x10}) {    // FNAME, FCOMMENT
      if ((flg & bit) != 0) {
        while (pos < src.length && src[pos] != 0) {
          pos++;
        }
        pos++;
      }
    }
    if ((flg & 0x02) != 0) {      // FHCRC
      pos += 2;
    }
    if (pos + 8 >= src.length) {
      throw new CodecException("gzip 머리가 잘렸다");
    }
    byte[] body = Arrays.copyOfRange(src, pos, src.length - 8);
    byte[] out = Inflate.inflateRaw(body);
    int base = src.length - 8;
    int crc = 0;
    int size = 0;
    for (int i = 3; i >= 0; i--) {
      crc = (crc << 8) | (src[base + i] & 0xFF);
      size = (size << 8) | (src[base + 4 + i] & 0xFF);
    }
    if (Checksums.crc32(out) != crc) {
      throw new CodecException("CRC-32 가 다르다");
    }
    if (out.length != size) {
      throw new CodecException("ISIZE 가 푼 길이와 다르다");
    }
    return out;
  }
}

File: compress/src/py/compresslib/containers.py
# -*- coding: utf-8 -*-
"""zlib 과 gzip 컨테이너 — SPEC §10.8 (RFC 1950, RFC 1952).

DEFLATE 스트림 자체에는 "무엇이었는지" 도 "제대로 풀렸는지" 도 없다.
그 둘을 얹은 것이 컨테이너다. zlib 은 2바이트 머리 + Adler-32,
gzip 은 10바이트 머리 + CRC-32 + 원본 길이.

gzip 머리의 MTIME 을 **0 으로 못 박는다.** 진짜 gzip 은 파일의 수정
시각을 적어서 같은 입력에 같은 바이트가 안 나온다 — 재현이 안 된다.
우리는 make record 를 세 번 돌려 md5 가 같아야 하므로 0 을 쓴다. 이 차이
자체가 9부의 이야깃거리다.
"""
from compresslib import checksums, deflate, inflate

# CMF = 0x78 (CM 8 = deflate, CINFO 7 = 32 KiB 창)
# FLG = 0x9C — FDICT 0, FLEVEL 2, 그리고 (CMF<<8 | FLG) 가 31 의 배수
ZLIB_CMF = 0x78
ZLIB_FLG = 0x9C

GZIP_MAGIC = b'\x1f\x8b'
GZIP_DEFLATE = 8
GZIP_OS_UNKNOWN = 255


def zlib_compress(src):
    head = bytes([ZLIB_CMF, ZLIB_FLG])
    tail = checksums.adler32(src).to_bytes(4, 'big')
    return head + deflate.deflate_raw(src) + tail


def zlib_decompress(src):
    if len(src) < 6:
        raise ValueError('zlib 스트림이 너무 짧다')
    cmf, flg = src[0], src[1]
    if (cmf & 0x0F) != 8:
        raise ValueError('zlib CM 이 8 이 아니다: %d' % (cmf & 0x0F))
    if ((cmf << 8) | flg) % 31:
        raise ValueError('zlib 머리의 검사식이 31 로 안 나눠진다')
    if flg & 0x20:
        raise ValueError('미리 정한 사전(FDICT)은 지원하지 않는다')
    out = inflate.inflate_raw(src[2:-4])
    want = int.from_bytes(src[-4:], 'big')
    if checksums.adler32(out) != want:
        raise ValueError('Adler-32 가 다르다')
    return out


def gzip_compress(src):
    head = GZIP_MAGIC + bytes([GZIP_DEFLATE, 0,
                               0, 0, 0, 0,        # MTIME = 0 (재현성)
                               0, GZIP_OS_UNKNOWN])
    tail = (checksums.crc32(src).to_bytes(4, 'little')
            + (len(src) & 0xFFFFFFFF).to_bytes(4, 'little'))
    return head + deflate.deflate_raw(src) + tail


def gzip_decompress(src):
    if len(src) < 18:
        raise ValueError('gzip 스트림이 너무 짧다')
    if src[:2] != GZIP_MAGIC:
        raise ValueError('gzip 매직이 아니다')
    if src[2] != GZIP_DEFLATE:
        raise ValueError('gzip CM 이 8 이 아니다: %d' % src[2])
    flg = src[3]
    pos = 10
    if flg & 0x04:                                   # FEXTRA
        n = src[pos] | (src[pos + 1] << 8)
        pos += 2 + n
    for bit in (0x08, 0x10):                         # FNAME, FCOMMENT
        if flg & bit:
            while pos < len(src) and src[pos]:
                pos += 1
            pos += 1
    if flg & 0x02:                                   # FHCRC
        pos += 2
    if pos >= len(src) - 8:
        raise ValueError('gzip 머리가 잘렸다')
    out = inflate.inflate_raw(src[pos:-8])
    crc = int.from_bytes(src[-8:-4], 'little')
    size = int.from_bytes(src[-4:], 'little')
    if checksums.crc32(out) != crc:
        raise ValueError('CRC-32 가 다르다')
    if (len(out) & 0xFFFFFFFF) != size:
        raise ValueError('ISIZE 가 푼 길이와 다르다')
    return out
